### DISCUSSION TITLE
ml-kem vs kyber & ml-dsa vs dilithium

### DISCUSSION THREAD

**Discussion by abhi-dev-engg** [open-quantum-safe/liboqs#1759] [General]
Hello Team,

I noticed a folder of ml-kem in kem along with kyber.
i assumed that ml-kem would be completely based on NIST FIPS203 specification, however i notice differences in that too wrt specification(derand etc..)

So, whats ml-kem exactly based on ? and which is the closest implementation to NIST specification then ?
and is ml-kem-ipd same as ml-kem ? as for tests both are considered same.

**Comment by baentsch** [comment]
> which is the closest implementation to NIST specification then 

Those are/should be "ml-kem" and "ml-dsa". Please document any functional (KAT?) differences you can see as [Bugs in liboqs](https://github.com/open-quantum-safe/liboqs/issues/new?assignees=&labels=&projects=&template=bug_report.md&title=).

> is ml-kem-ipd same as ml-kem ?

Yes -- for now: The name "ml-kem" will remain the same until the algorithm is finally standardized (but the underlying implementation may change). Please see [the documentation](https://github.com/open-quantum-safe/liboqs?tab=readme-ov-file#supported-algorithms).

**Comment by abhi-dev-engg** [comment]
@baentsch thanks for the reply. got it now !

Would it be a good idea to set DILITHIUM_MODE = 2/3/5 in https://github.com/open-quantum-safe/liboqs/blob/main/src/sig/ml_dsa/pqcrystals-dilithium-standard_ml-dsa-65-ipd_ref/config.h#L10(actual value supplied from Cmake compile options). Currently its set to 2 in all files.

Also(being greedy :) ), i noticed you as a contributor in kyber github repo. Is it possible to check https://github.com/pq-crystals/kyber/issues/74

Thank you again and sorry for the multiple questions.

Regards.

**Reply by baentsch** [reply]
> Currently its set to 2 in all files.

This indeed seems weird. But this code is/should be a direct copy of the upstream code base. Any feedback, @bhess? But @yogappu, for `liboqs` this is no problem as this define is overruled by the "compile_opts" setting e.g., [here](https://github.com/open-quantum-safe/liboqs/blob/7b6d9f3326295fc80ea0c9026f3dd9d57f8436de/src/sig/ml_dsa/CMakeLists.txt#L31). Maybe easier done than changing the upstream code.

> Is it possible to check https://github.com/pq-crystals/kyber/issues/74

I'm not sure who's looking after that repo; I do not (and my contribution was a "historical mistake" :-)

**Reply by bhess** [reply]
Yes, it's a direct copy from upstream where the default is 2 if `DILITHIUM_MODE` is not supplied. We set `DILITHIUM_MODE` with compile opts as @baentsch correctly says.

Re. https://github.com/pq-crystals/kyber/issues/74, I'll comment upstream.

**Comment by abhi-dev-engg** [comment]
Closing as queries answered
